File: route-service/utils/way_point_helper.py
from constants.map_type import SportMapType
from dto.supbase.response.way_point import WayPoint


class WayPointHelper:
    @staticmethod
    def get_rdp_epsilon(map_type: SportMapType) -> float:
        if map_type == SportMapType.WALKING:
            return 0.00005
        return 0.001

    @staticmethod
    def get_simplification_step(map_type: SportMapType) -> int:
        if map_type == SportMapType.WALKING:
            return 5
        return 10

    @staticmethod
    def filter_points(points: list[WayPoint]) -> list[WayPoint]:
        filtered = WayPointHelper._remove_none(points)
        filtered = WayPointHelper._filter_by_local_repetition(filtered)
        filtered = WayPointHelper._keep_run_boundaries(filtered)
        return filtered

    @staticmethod
    def _remove_none(points: list[WayPoint]) -> list[WayPoint]:
        """Remove None entries from the list."""
        return [p for p in points if p is not None]

    @staticmethod
    def _filter_by_local_repetition(points: list[WayPoint], window_size: int = 2, min_count: int = 3) -> list[WayPoint]:
        """
        Keep points that appear at least `min_count` times within a sliding window around each point.
        """
        n = len(points)
        if n < window_size * 2 + 1:
            return points

        result = []

        for i in range(window_size):
            result.append(points[i])


        for i in range(window_size, n - window_size):
            window = points[i - window_size: i + window_size + 1]
            name = points[i].name
            count = sum(1 for p in window if p.name == name)
            if count >= min_count:
                result.append(points[i])

        for i in range(n - window_size, n):
            result.append(points[i])

        return result

    @staticmethod
    def _keep_run_boundaries(points: list[WayPoint]) -> list[WayPoint]:
        """
        Keep only the first and last point of each run of consecutive points with the same name.
        """
        if not points:
            return []

        result = []
        i = 0
        while i < len(points):
            start = i
            current_name = points[i].name
            while i + 1 < len(points) and points[i + 1].name == current_name:
                i += 1
            end = i
            result.append(points[start])
            if end != start:
                result.append(points[end])
            i += 1

        return result

File: core-service/src/main/java/com/stride/tracking/coreservice/utils/WayPointHelper.java
package com.stride.tracking.coreservice.utils;

import com.stride.tracking.core.dto.sport.SportMapType;
import com.stride.tracking.core.dto.supabase.response.WayPoint;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class WayPointHelper {
    private WayPointHelper() {
    }

    public static double getRdpEpsilon(SportMapType mapType) {
        if (mapType == SportMapType.WALKING) {
            return 0.00005;
        }
        return 0.001;
    }

    public static int getSimplificationStep(SportMapType mapType) {
        if (mapType == SportMapType.WALKING) {
            return 5;
        }
        return 10;
    }

    public static List<WayPoint> filterPoints(List<WayPoint> points) {
        List<WayPoint> filtered = removeNone(points);
        filtered = filterByLocalRepetition(filtered, 2, 3);
        filtered = keepRunBoundaries(filtered);
        return filtered;
    }

    private static List<WayPoint> removeNone(List<WayPoint> points) {
        List<WayPoint> result = new ArrayList<>();
        for (WayPoint point : points) {
            if (point != null) {
                result.add(point);
            }
        }
        return result;
    }

    private static List<WayPoint> filterByLocalRepetition(List<WayPoint> points, int windowSize, int minCount) {
        int n = points.size();
        if (n < windowSize * 2 + 1) {
            return new ArrayList<>(points);
        }

        List<WayPoint> result = new ArrayList<>();

        for (int i = 0; i < windowSize; i++) {
            result.add(points.get(i));
        }

        for (int i = windowSize; i < n - windowSize; i++) {
            List<WayPoint> window = points.subList(i - windowSize, i + windowSize + 1);
            String name = points.get(i).getName();
            long count = window.stream().filter(p -> Objects.equals(p.getName(), name)).count();
            if (count >= minCount) {
                result.add(points.get(i));
            }
        }

        for (int i = n - windowSize; i < n; i++) {
            result.add(points.get(i));
        }

        return result;
    }

    private static List<WayPoint> keepRunBoundaries(List<WayPoint> points) {
        List<WayPoint> result = new ArrayList<>();
        if (points == null || points.isEmpty()) {
            return result;
        }

        int i = 0;
        while (i < points.size()) {
            int start = i;
            String currentName = points.get(i).getName();
            while (i + 1 < points.size() && Objects.equals(points.get(i + 1).getName(), currentName)) {
                i++;
            }
            int end = i;
            result.add(points.get(start));
            if (end != start) {
                result.add(points.get(end));
            }
            i++;
        }

        return result;
    }
}
